File: 51.n-皇后.go
/*
 * @lc app=leetcode.cn id=51 lang=golang
 *
 * [51] N 皇后
 */

// @lc code=start
func solveNQueens(n int) [][]string {
	// 恰好每行每列有一个
	// r 表示行号，c 表示列号
	// r+c，假设要在 (2, 0) 放置，则不能在 r+c=2 的其他地方放置，判断右上方的皇后
	// r-c，假设要在 (2, 0) 放置，则不能在 r-c=-1 的其他地方放置，判断左上方的皇后
	var boards [][]string
	var board []string
	// r 当前枚举的行号，s 表示剩余可枚举的列号
	var dfs func(row int)
	var checker func(row, col int) bool

	cols := make([]int, n)
	viewed := make([]bool, n)

	checker = func(row, col int) bool {
		for R := 0; R < row; R++ {
			C := cols[R]
			if row+col == R+C || row-col == R-C {
				return false
			}
		}

		return true
	}

	dfs = func(row int) {
		if row == n {
			board = make([]string, n)
			for i, c := range cols {
				board[i] = strings.Repeat(".", c) + "Q" + strings.Repeat(".", n-1-c)
			}
			boards = append(boards, append([]string{}, board...))
			return
		}

		for col := 0; col < n; col++ {
			if !viewed[col] && checker(row, col) {
				cols[row] = col
				viewed[col] = true
				dfs(row + 1)
				viewed[col] = false
				cols[row] = 0
			}
		}
	}

	dfs(0)

	return boards
}

// @lc code=end

File: 51.n-皇后.py
#
# @lc app=leetcode.cn id=51 lang=python
#
# [51] N 皇后
#

# @lc code=start
class Solution(object):
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        row_cols = [-1] * n
        cols = [False] * n
        paths = []
        def check(row, col):
            for R in range(0, row):
                C = row_cols[R]
                if row+col == R+C or row-col == R-C:
                    return False
            return True

        def dfs(row):
            if row > n:
                return
            if row == n:
                path = []
                for i in range(0, n):
                    col = row_cols[i]
                    path.append("."*col+"Q"+"."*(n-1-col))
                paths.append(path)
            for col in range(0, n):
                if cols[col] == False and check(row, col):
                    cols[col] = True
                    row_cols[row] = col
                    dfs(row+1)
                    row_cols[row] = -1
                    cols[col] = False
        dfs(0)
        return paths



# @lc code=end

